### PR TITLE
Trace output

### PR DESCRIPTION
**Pull request by fantypants**
Trace-Output:

- Adds option of ```:trace```
- Updated ReadMe to include the new options
- Allows you to see the executed instructions & the accumulated gas cost in the system, handy for debugging instruction sets and gas cost

### CODE REVIEW

File: lib/execution/executor.ex [hunk 1]
```diff
@@ -55,15 +55,18 @@ defmodule WaspVM.Executor do
...
+  def execute(frame, vm, gas, stack, total_instr, gas_limit, trace, next_instr \\ 0)
+  def execute(_frame, vm, gas, stack, _total, gas_limit, trace, _next) when gas_limit != :infinity and gas > gas_limit, do: IEx.pry #{:error, :reached_gas_limit}
+  def execute(_frame, vm, gas, stack, total_instr, _gas_limit, trace, next_instr) when next_instr >= total_instr or next_instr < 0, do: {vm, gas, stack}
+  def execute(frame, vm, gas, stack, total_instr, gas_limit, trace, next_instr) do
     %{^next_instr => instr} = frame.instructions
-
     {{frame, vm, next_instr}, gas, stack} = instruction(instr, frame, vm, gas, stack, next_instr)
 
-    execute(frame, vm, gas, stack, total_instr, gas_limit, next_instr + 1)
+    if trace == true do
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: We should never be comparing to a boolean in any expression, since the expression already gets evaluated to a boolean. This should be rewritten as
```elixir
if trace, do: write_to_file(instr, gas)
```

File: lib/execution/executor.ex [hunk 2]
```diff
@@ -894,4 +897,16 @@ defmodule WaspVM.Executor do
 
     Integer.undigits(zero_leading_map ++ bin, 2)
   end
+
+  defp create_entry(instruction) when is_tuple(instruction) == false, do: "#{instruction}"
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Same as above, don't compare against a boolean. We also don't need to interpolate into a string if we're not adding anything with it -- if we need it to be a string, we should call `Kernel.to_string\1`. 

This should be rewritten as 
```elixir
defp create_entry(instruction) when not is_tuple(instruction), do: instruction
```

File: lib/execution/executor.ex [hunk 3]
```diff
@@ -894,4 +897,16 @@ defmodule WaspVM.Executor do
...
+  defp create_entry({instruction, variable}), do: "#{instruction}"
+
+  defp write_to_file(instruction, gas) do
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Maybe we shouldn't default to writing this to a file, and instead allow the user to decide what they want to do with this output? It might make sense to create a file-like string (newline delimited) and just spit it out as a return value, at which point the developer can decide if they want to write it to a file.

As an example, I know I'd only use the trace output for debugging, so I'd just want it written to my terminal rather than a file, but other people might want to keep logs, so they'd be able to take the value passed back by trace and write it to a log file

File: lib/wasp_vm.ex [hunk 4]
```diff
@@ -160,19 +160,19 @@ defmodule WaspVM do
     {reply, vm} =
       case func_addr do
         :not_found -> {{:error, :no_exported_function, fname}, vm}
-        addr -> execute_func(vm, addr, args, opts[:gas_limit])
+        addr -> execute_func(vm, addr, args, [opts[:gas_limit], opts[:trace]])
```
Comment (alexdovzhanyn) [inline lib/wasp_vm.ex]: Lets just pass the entire opts here since at this point that's all we're using it for. We can split it out later if we need to.

File: text.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+new content!
```
Comment (alexdovzhanyn) [inline text.txt]: Is this file necessary? Doesn't seem to be referenced anywhere

File: README.md [hunk 6]
```diff
@@ -9,13 +9,31 @@ the [Elixium Network](https://www.elixiumnetwork.org)
...
 ```elixir
 {:ok, ref} = WaspVM.start() # Start WaspVM
 WaspVM.load_file(ref, "path/to/wasm/file.wasm") # Load a module
 WaspVM.execute(ref, "some_exported_function") # Call a function
 # => {:ok, total_gas_cost, :function_return_value}
 ```
 
+# Options
+For Gas Limit use:
+``` elixir
```
Comment (alexdovzhanyn) [inline README.md]: These options should also be added to the `execute/4` function docs so they show up on the hexdocs

File: lib/execution/executor.ex (hunk 2 above)
Comment (fantypants) [reply lib/execution/executor.ex]: Fixed

File: lib/execution/executor.ex (hunk 1 above)
Comment (fantypants) [reply lib/execution/executor.ex]: Fixed

File: lib/wasp_vm.ex (hunk 4 above)
Comment (fantypants) [reply lib/wasp_vm.ex]: Fixed

File: README.md (hunk 6 above)
Comment (fantypants) [reply README.md]: Adding Now

File: lib/execution/executor.ex (hunk 3 above)
Comment (fantypants) [reply lib/execution/executor.ex]: Resolved

File: text.txt (hunk 5 above)
Comment (fantypants) [reply text.txt]: Fixed

File: lib/execution/executor.ex [hunk 7]
```diff
@@ -894,4 +898,18 @@ defmodule WaspVM.Executor do
...
   end
+
+  defp create_entry(instruction) when not is_tuple(instruction), do: Kernel.to_string(instruction)
+  defp create_entry({instruction, variable}), do: Kernel.to_string(instruction)
+
+  defp write_to_file(instruction, gas) do
+    file =
+      Path.expand('./trace_log.log')
+      |> Path.absname
+      |> File.write(create_entry(instruction) <> " " <> "#{gas}" <> "\n", [:append])
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: This should be rewritten as 
```elixir
File.write("#{create_entry(instruction} #{gas}\n", [:append)
```

File: test/executor_test.exs [hunk 8]
```diff
@@ -239,7 +239,7 @@ defmodule WaspVM.ExecutorTest do
       {:ok, pid} = WaspVM.start()
       WaspVM.load_file(pid, "test/fixtures/wasm/nested_func_call.wasm")
 
-      {:ok, gas, result} =  WaspVM.execute(pid, "nested_func_call", [0, 32])
+      {:ok, gas, result} =  WaspVM.execute(pid, "nested_func_call", [0, 32], [gas_limit: :infinity])
```
Comment (alexdovzhanyn) [inline test/executor_test.exs]: Gas limit defaults to `:infinity` if it isn't passed IIRC, so this doesn't need to be here

File: test/program_test.exs [hunk 9]
```diff
@@ -14,4 +14,13 @@ defmodule WaspVM.ProgramTest do
     assert result_2 == 2
   end
 
+  test "Trace Works" do
+    {:ok, pid} = WaspVM.start()
+    WaspVM.load_file(pid, "test/fixtures/wasm/int_div.wasm")
+    {status, gas, result} = WaspVM.execute(pid, "main", [-4], [trace: true])
+
+
+    assert result == -2
```
Comment (alexdovzhanyn) [inline test/program_test.exs]: We should actually read the trace file in and make sure things got written to it correctly (and don't forget to have the file deleted automatically after the test finishes running)

File: trace_log.log [hunk 10]
```diff
@@ -0,0 +1,49 @@
+
```
Comment (alexdovzhanyn) [inline trace_log.log]: Let's gitignore this file (and also rename it to trace.log)

File: lib/execution/executor.ex (hunk 7 above)
Comment (fantypants) [reply lib/execution/executor.ex]: Fixed

File: test/executor_test.exs (hunk 8 above)
Comment (fantypants) [reply test/executor_test.exs]: Fixed, Removed

File: test/program_test.exs (hunk 9 above)
Comment (fantypants) [reply test/program_test.exs]: Appended test to read the outputted file, check for > 0 string size, and delete test file on close

File: trace_log.log (hunk 10 above)
Comment (fantypants) [reply trace_log.log]: added git ignore to all .log files

**Comment by alexdovzhanyn** [comment]
This build is failing: https://travis-ci.org/ElixiumNetwork/WaspVM/builds/482724486#L480, its not compiling

**Comment by fantypants** [comment]
Fixed Compilation Issues, It had left git tags again and rearranged some stuff.

**Comment by alexdovzhanyn** [comment]
We're almost there with this PR. There are still some things that seem to have slipped through the cracks -- upon testing the PR it seems like the trace option crashes the VM when using if statements:

```
** (FunctionClauseError) no function clause matching in WaspVM.Executor.create_entry/1
(wasp_vm) lib/execution/executor.ex:1082: WaspVM.Executor.create_entry({:if, :no_res, :none, 8})
```

This won't work because there isn't a matching create_entry function clause. We need to add a case to match this (and we need a default case that emits a warning to IO and then skips logging it to the file.

**Comment by alexdovzhanyn** [comment]
> We're almost there with this PR. There are still some things that seem to have slipped through the cracks -- upon testing the PR it seems like the trace option crashes the VM when using if statements:
> 
> ```
> ** (FunctionClauseError) no function clause matching in WaspVM.Executor.create_entry/1
> (wasp_vm) lib/execution/executor.ex:1082: WaspVM.Executor.create_entry({:if, :no_res, :none, 8})
> ```
> This won't work because there isn't a matching create_entry function clause. We need to add a case to match this (and we need a default case that emits a warning to IO and then skips logging it to the file.

I've fixed this in the latest commit to this PR
